BedrockServer.cpp

```diff
@@ -1682,28 +1502,34 @@ void BedrockServer::postPoll(fd_map& fdm, uint64_t& nextActivity) {
         }
     }
 
-    // If we've been told to start shutting down, we'll set the lastChance timer.
+    // If we've been told to start shutting down, we'll set the _lastChance timer.
     if (_shutdownState.load() == START_SHUTDOWN) {
-        if (!lastChance) {
-            lastChance = STimeNow() + 5 * 1'000'000; // 5 seconds from now.
+        if (!_lastChance) {
+            _lastChance = STimeNow() + 5 * 1'000'000; // 5 seconds from now.
         }
         // If we've run out of sockets or hit our timeout, we'll increment _shutdownState.
-        if (socketList.empty() || _gracefulShutdownTimeout.ringing()) {
-            lastChance = 0;
+        if ((socketList.empty() && !_outstandingSocketThreads) || _gracefulShutdownTimeout.ringing()) {
+            _lastChance = 0;
 
             // We empty the socket list here, we will no longer allow new requests to come in, as the sync node can
             // shutdown any time after here, and we'll have no way to handle new requests.
             if (socketList.size()) {
-                SAUTOLOCK(_socketIDMutex);
                 SINFO("Killing " << socketList.size() << " remaining sockets at graceful shutdown timeout.");
                 while(socketList.size()) {
                     auto s = socketList.front();
-                    _socketIDMap.erase(s->id);
                     closeSocket(s);
                 }
             }
+
             _shutdownState.store(CLIENTS_RESPONDED);
         }
+        if (_outstandingSocketThreads) {
+            SINFO("Have " << _outstandingSocketThreads << " socket threads to close.");
+        }
+        size_t count = BedrockCommand::getCommandCount();
```

Comment:
Because this is more likely to be correct. Incrementing a count in a constructor and decrementing it in the destructor is a pretty straightforward way to count the number of existing commands.

BedrockServer hands commands all over the place (from workers to the sync thread, to SQLiteNode, to plugins, etc), and keeping track of all those is easy to break. The whole point of this check is just to know if we've lost any commands anywhere, we want the simplest way to detect that.